First pass at similar frame deletion

RaphiePS opened the pull request:
Heyo! Was completely blown away by _Triple-Chaser_ at the Biennial, and would love to help out however I can (I'll shoot you an email as well). This is an initial solution to Issue #9.

### Algorithm

The similar-frame heuristic here is pretty basic, and could eventually be swapped out for something much more clever like optical flow. But it seems to work pretty well: it keeps only one frame for static image videos ("I love techno") and 98-100% of frames for non-static videos. Here's how it works:
1) Convert frame to grayscale, and scale the pixels from 0-255 to 0-1.
2) Subtract the previously output frame, and take the L2 norm. Divide by number of pixels to make it invariant to image size. This gives us a "change score."
3) Compare the score to `change_threshold`, and only save the frame if passes. I've set the threshold to `1e-5`, but you can totally play with this value - lower means it'll keep more frames (0 for keeping every frame), and higher means it'll more aggressively prune similar frames.

### JSON config parameters
- `method` enables or disables this new frame-output algorithm. `ffmpeg` will use the existing system, while `opencv` will use the new one.
- `change_threshold` sets the threshold as above
- `fps` is the output frames per second, now configurable in JSON
- `sequential`: I wasn't quite sure what you meant by keeping frame numbers intact. For example, say we have frames 1 through 5, and we decide to remove frame 4. If `sequential` is true, the output frames will be named: `1, 2, 3, 4`. If `sequential` is false, then we'll keep the original names and output `1, 2, 3, 5`. 

### A few notes
1) This comes with a noticeable performance penalty. I ran both the original `ffmpeg` method and the new `opencv` method on the YouTube videos downloaded in the first selection stage. On my Macbook Pro (2.9 GHz i7, 16 GB RAM), the original ffmpeg takes **30.4s**, whereas the new opencv takes **47.8s**. Hopefully this is a worthy tradeoff for the space savings.
2) You'll need to rebuild the Docker image to grab the new `imutils` pip dependency. 

Happy to tweak this if there are any issues. Thanks again for all your work!

RaphiePS committed:
First pass at frame deletion

RaphiePS committed:
Update defaults

breezykermo commented:
This looks great! The performance penalty is admissible IMO, as in addition to space-saving in the immediate, this feature also saves compute time down the line in our workflow (potentially drastically), as the reason for the `frames` analyser is to be able to apply image CV classifiers in the next pass, and analyse the contents of the video.

This workflow (frames --> image CV classifier) is also the cause of my cryptic "keeping frames intact" (sorry!). When extracting frames from a video, we want to know which second in the video the frame corresponds to. When frames was just extracting 1fps, this is easy to track by the number of the frame-- but when dropping frames it's important for this workflow to keep that information in some way in the frame name.

We have a frontend for mtriage that visualises the results of a youtube --> frames --> pytorch workflow. It's not yet OSS (hopefully soon!), but this screenshot might be helpful to consider what I mean about retaining the information. Each black cell represents a video's timeline, and the red flecks indicate where an image CV classifier has returned a positive result:

<img width="1893" alt="Screen Shot 2019-05-27 at 10 19 58 AM" src="https://user-images.githubusercontent.com/16495856/58410262-bc43eb80-8069-11e9-930e-447737333021.png">

Does this make sense? I believe that the `sequential` optionality you have in this PR addresses this. For other workflows this "keeping intact" may not be relevant, so I like having the option to elide this meta information in there as well.

RaphiePS committed:
Make sequential=false the default

RaphiePS committed:
Make scripts executable

RaphiePS committed:
Better framerate sampling to avoid timestamp drift

RaphiePS commented:
Makes total sense - I've updated the defaults to set `sequential` to false, keeping those filenames intact. To make sure this is working properly, I made a quick [test video](https://www.dropbox.com/s/6ea37en9hbv0b3h/test.mp4?dl=0) with 5 frames at 1fps. Frames 0, 1, 2, and 4 are unique (A, B, C, D) whereas frame 3 is a duplicate of frame 2 (C). With the new default `sequential=false`, the analyser will now output `0000.bmp  0001.bmp  0002.bmp  0004.bmp`. I think this is the correct behavior.

I also realized I had a subtle bug. It turns out, believe it or not, that some YouTube videos are encoded at wacky values like *12.11fps*. Why they use fractional FPS values I'll never know, but this means I need more nuanced FPS handling. The old technique would pick every `nth` frame, where `n = int(video_fps / desired_fps)`. In the case of 12.11fps videos, this would choose every 12th frame, which will make the timestamps drift over time. Now fixed!

breezykermo commented:
Fantastic thanks, this works great! I'm going to leave #9 open, but this is an excellent first pass.

breezykermo commented on scripts/2.frames.sh:
```diff
@@ -7,7 +7,21 @@ fi
+	--argjson fps "$FPS" \
+	'{method: $m, change_threshold: $ct, sequential: $seq, fps: $fps}'
+)
+
```
My only comment would be that these parameters could be better structured/documented-- but that's not to do with this PR, it's on us c.f. #25
